### PR TITLE
Add extending kubernetes blog post

### PR DESCRIPTION
**Pull request by holger-wg2**

### CODE REVIEW

File: pages/blog/_posts/2020-02-20-Extending-Kubernetes-Static-IP-Routing.md [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+title: "Extending Kubernetes for our needs"
+date: 2020-01-28
+tags: infrastructure kubernetes operator operator-framework networking aws
+author: <a href="https://www.linkedin.com/in/hihrig/">Holger Ihrig</a>
+---
+
+We are using Kubernetes as our cluster scheduler and this serves us well. However we have a 
+few cases where we need to do some additional work.
+
+One case for example is that we have to use static-IPs for some of our services to connect to
```
Comment (annaken) [inline pages/blog/_posts/2020-02-20-Extending-Kubernetes-Static-IP-Routing.md]: You don't need the hyphen in static IPs

File: pages/blog/_posts/2020-02-20-Extending-Kubernetes-Static-IP-Routing.md [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+Telecom companies as they expect a single IP address to bind to. This address needs to be the same and
```
Comment (annaken) [inline pages/blog/_posts/2020-02-20-Extending-Kubernetes-Static-IP-Routing.md]: This address needs to be the same 
-> the same as what? or do you mean static?

File: pages/blog/_posts/2020-02-20-Extending-Kubernetes-Static-IP-Routing.md (hunk 1 above)
Comment (holger-wg2) [reply pages/blog/_posts/2020-02-20-Extending-Kubernetes-Static-IP-Routing.md]: Fixed

File: pages/blog/_posts/2020-02-20-Extending-Kubernetes-Static-IP-Routing.md (hunk 2 above)
Comment (holger-wg2) [reply pages/blog/_posts/2020-02-20-Extending-Kubernetes-Static-IP-Routing.md]: changed this to "This address needs to be fixed"

File: pages/blog/_posts/2020-02-20-Extending-Kubernetes-Static-IP-Routing.md [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+DNS records are not accepted either. We are running in AWS as well, so the reader might ask why we are not
+using Elastic IPs and add them to the services? Good idea, but Telecom operators will not whitelist
+an Elastic IP for you as there are no guarantees that it will belong to your Infrastructure forever.
+
+We are a member of RIPE and do have a small subnet block for our own use, so we thought we could make use of 
+that. As we own the block and AWS supports BYOIR (Bring your own IP-range), we created a special subnet with
+some Kubernetes Nodes in it. This was not enough to make this work since we depended on a service always having
```
Comment (annaken) [inline pages/blog/_posts/2020-02-20-Extending-Kubernetes-Static-IP-Routing.md]: (n)odes

File: pages/blog/_posts/2020-02-20-Extending-Kubernetes-Static-IP-Routing.md [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+author: <a href="https://www.linkedin.com/in/hihrig/">Holger Ihrig</a>
+---
+
+We are using Kubernetes as our cluster scheduler and this serves us well. However we have a 
+few cases where we need to do some additional work.
+
+One case for example is that we have to use static-IPs for some of our services to connect to
+Telecom companies as they expect a single IP address to bind to. This address needs to be the same and
+DNS records are not accepted either. We are running in AWS as well, so the reader might ask why we are not
+using Elastic IPs and add them to the services? Good idea, but Telecom operators will not whitelist
```
Comment (annaken) [inline pages/blog/_posts/2020-02-20-Extending-Kubernetes-Static-IP-Routing.md]: why we are not using Elastic IPs and add them to the services?
-> why are we not using Elastic IPs and adding them to the services?

File: pages/blog/_posts/2020-02-20-Extending-Kubernetes-Static-IP-Routing.md [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+node on how good they meet the requirements. The Node that fits best, gets chosen.
+
+All of that are things the scheduler will do for you automatically, however it is also possible to give the cluster a 
+`KubeSchedulerConfiguration` object that will tell the scheduler to also reach out to a service for additional point scoring.
+The SchedulerConfiguration is a JSON file and for further explanations, please have a look at this 
+excellent [blog post](https://developer.ibm.com/articles/creating-a-custom-kube-scheduler/).
+
+In our case, we could have written a service that checks which IP Addresses are assigned to the Nodes and moved the Pods
+onto those nodes. This would have required us to make sure that those Nodes had all needed IP-Addresses all the time.
+That sounded not very enticing when doing cluster upgrades as it would have needed to be at least a semi-manual process.
```
Comment (annaken) [inline pages/blog/_posts/2020-02-20-Extending-Kubernetes-Static-IP-Routing.md]: I'm left a bit confused as to whether you implemented a scheduler extender or not.

File: pages/blog/_posts/2020-02-20-Extending-Kubernetes-Static-IP-Routing.md [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+Operators usually consists at least out of a CRD(CustomResourceDefinition), an Event Listener and a Reconciliation Loop.
+The CRD defines the object that shall be created, e.g. a Redis deployment. The whole description on what this deployment
+needs to look like and all its abilities need to be defined in the CRD. The Operator will create an Event Listener for that
+CRD as the primary resource and additional event listeners for the secondary resource (most likely pods in this example).
+The Event Listener will let the Reconciliation Loop know once a CRUD operation has been requested on either the primary or
+secondary resource. The Loop will then try to bring the CRD into the desired state, depending on what operation has been
+requested. So in the Redis example it will either create the pods, update the pods, in the case of an upgrade it might blue/green
+the pods or delete the pods. Basically things a human operator would do in this case, just in programmatic form, taking it
+from a declarative form into existing resources.
```
Comment (annaken) [inline pages/blog/_posts/2020-02-20-Extending-Kubernetes-Static-IP-Routing.md]: There's a lot of information in this paragraph - maybe split it up a bit?

File: pages/blog/_posts/2020-02-20-Extending-Kubernetes-Static-IP-Routing.md (hunk 5 above)
Comment (holger-wg2) [reply pages/blog/_posts/2020-02-20-Extending-Kubernetes-Static-IP-Routing.md]: Will make that more clear

File: pages/blog/_posts/2020-02-20-Extending-Kubernetes-Static-IP-Routing.md (hunk 6 above)
Comment (holger-wg2) [reply pages/blog/_posts/2020-02-20-Extending-Kubernetes-Static-IP-Routing.md]: I tried splitting it up, but did not succeed ... did sprinkle a little bit of highlighting on several words to make it easier to follow though

File: pages/blog/_posts/2020-02-20-Extending-Kubernetes-Static-IP-Routing.md [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+---
+
+We are using Kubernetes as our cluster scheduler and this serves us well. However we have a 
+few cases where we need to do some additional work.
+
+One case for example is that we have to use static IPs for some of our services to connect to
+Telecom companies as they expect a single IP address to bind to. This address needs to be fixed and
+DNS records are not accepted either. We are running in AWS as well, so the reader might ask why are we not
+using Elastic IPs and adding them to the services? Good idea, but Telecom operators will not whitelist
+an Elastic IP for you as there are no guarantees that it will belong to your company infrastructure forever.
```
Comment (mtl-wgtwo) [inline pages/blog/_posts/2020-02-20-Extending-Kubernetes-Static-IP-Routing.md]: Also can't use Elastic IPs in a private subnet over VPN or direct connect.

File: pages/blog/_posts/2020-02-20-Extending-Kubernetes-Static-IP-Routing.md [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+
+There are more ways to extend Kubernetes, but these two ways will be the one we shall look at. Just for completeness
+you can as well also add another scheduler or change kubernetes itself. However these possibilities have some serious
+downsides.
+
+### Adding a scheduler extender
+The kubernetes scheduler checks for certain requirements before it schedules a pod onto a node. Some of these requirements
+are hard requirements, like cpu, memory and number of pods. Other requirements are more soft, like if the pods are allowed to
+be packed together or in which AZ they are going to run. All of those requirements are collected and points given to each
+node on how good they meet the requirements. The node that fits best, gets chosen.
```
Comment (mtl-wgtwo) [inline pages/blog/_posts/2020-02-20-Extending-Kubernetes-Static-IP-Routing.md]: good -> well

File: pages/blog/_posts/2020-02-20-Extending-Kubernetes-Static-IP-Routing.md [hunk 9]
```diff
@@ -0,0 +1,303 @@
...
+
+All of that are things the scheduler will do for you automatically, however it is also possible to give the cluster a
```
Comment (mtl-wgtwo) [inline pages/blog/_posts/2020-02-20-Extending-Kubernetes-Static-IP-Routing.md]: that -> these

File: pages/blog/_posts/2020-02-20-Extending-Kubernetes-Static-IP-Routing.md [hunk 10]
```diff
@@ -0,0 +1,303 @@
...
+So we decided against this approach.
+
+### Creating an operator
+An Operator/Controller on the other hand is a component observing resources and then try to create the declared resources.
+The difference between a controller and an operator is basically that an operator is handling the lifecycle of an
+application, whereas a controller may control specific resources that are not associated with a specific application.
+They both use the controller pattern though and both can be implemented with the same toolset, so for simplicity sake in the
+context of this article, we will consider them to be equal.
+
+Operators usually consists at least out of a **CRD(CustomResourceDefinition)**, an **event listener** and a **reconciliation loop**.
```
Comment (mtl-wgtwo) [inline pages/blog/_posts/2020-02-20-Extending-Kubernetes-Static-IP-Routing.md]: Consider links to the documentation for these bolded bits, if possible.

File: pages/blog/_posts/2020-02-20-Extending-Kubernetes-Static-IP-Routing.md [hunk 11]
```diff
@@ -0,0 +1,303 @@
...
+The *CRD* defines the object that shall be created, e.g. a Redis deployment. The whole description on what this deployment
```
Comment (mtl-wgtwo) [inline pages/blog/_posts/2020-02-20-Extending-Kubernetes-Static-IP-Routing.md]: Maybe point out that the CRD is creating a new type in Kubernetes and effectively implements a domain-specific language for the operator.

File: pages/blog/_posts/2020-02-20-Extending-Kubernetes-Static-IP-Routing.md [hunk 12]
```diff
@@ -0,0 +1,303 @@
...
+needs to look like and all its abilities need to be defined in the *CRD*. The operator will create an *event listener* for that
+*CRD* as the primary resource and additional *event listeners* for the secondary resource (most likely pods in this example).
+The *event listener* will let the *reconciliation loop* know once a CRUD operation has been requested on either the primary or
```
Comment (mtl-wgtwo) [inline pages/blog/_posts/2020-02-20-Extending-Kubernetes-Static-IP-Routing.md]: maybe link CRUD to https://en.wikipedia.org/wiki/Create,_read,_update_and_delete?

File: pages/blog/_posts/2020-02-20-Extending-Kubernetes-Static-IP-Routing.md (hunk 11 above)
Comment (mtl-wgtwo) [reply pages/blog/_posts/2020-02-20-Extending-Kubernetes-Static-IP-Routing.md]: You could also mention here that the new type is called a *kind* later

**Comment by jonny-wg2** [comment]
Very good read :D

File: pages/blog/_posts/2020-02-20-Extending-Kubernetes-Static-IP-Routing.md [hunk 13]
```diff
@@ -0,0 +1,303 @@
...
+- Pod has moved to another node
+- Pod has been deleted
+
+We will not go any further into the detail on what these parts are actually doing, as this is enough
+to actually give you an idea on how this can be accomplished.
+
+
+## Summary
+
+We have been looking at how to expand kubernetes to suit our needs better. Creating the Operator/Controller
```
Comment (mtl-wgtwo) [inline pages/blog/_posts/2020-02-20-Extending-Kubernetes-Static-IP-Routing.md]: s/kubernetes/Kubernetes/g

File: pages/blog/_posts/2020-02-20-Extending-Kubernetes-Static-IP-Routing.md (hunk 7 above)
Comment (holger-wg2) [reply pages/blog/_posts/2020-02-20-Extending-Kubernetes-Static-IP-Routing.md]: Added a note on that

File: pages/blog/_posts/2020-02-20-Extending-Kubernetes-Static-IP-Routing.md (hunk 8 above)
Comment (holger-wg2) [reply pages/blog/_posts/2020-02-20-Extending-Kubernetes-Static-IP-Routing.md]: done

File: pages/blog/_posts/2020-02-20-Extending-Kubernetes-Static-IP-Routing.md (hunk 9 above)
Comment (holger-wg2) [reply pages/blog/_posts/2020-02-20-Extending-Kubernetes-Static-IP-Routing.md]: done

File: pages/blog/_posts/2020-02-20-Extending-Kubernetes-Static-IP-Routing.md (hunk 13 above)
Comment (holger-wg2) [reply pages/blog/_posts/2020-02-20-Extending-Kubernetes-Static-IP-Routing.md]: done

File: pages/blog/_posts/2020-02-20-Extending-Kubernetes-Static-IP-Routing.md (hunk 12 above)
Comment (holger-wg2) [reply pages/blog/_posts/2020-02-20-Extending-Kubernetes-Static-IP-Routing.md]: done

**Pull request merged**